Add Stream ReadAtLeast and ReadExactly

eerhardt opened the pull request:
Adds methods to Stream to read at least a minimum amount of bytes, or a full buffer, of data from the stream.
ReadAtLeast allows for the caller to specify whether an exception should be thrown or not on the end of the stream.

Make use of the new methods where appropriate in net7.0.

Fix #16598

eerhardt committed:
Add Stream ReadAtLeast and ReadExactly

Adds methods to Stream to read at least a minimum amount of bytes, or a full buffer, of data from the stream.
ReadAtLeast allows for the caller to specify whether an exception should be thrown or not on the end of the stream.

Make use of the new methods where appropriate in net7.0.

Fix #16598

campersau commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -332,6 +332,48 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
+        {
+            ValidateReadAtLeastArguments(buffer.Length, minimumBytes);
+
+            return ReadAtLeastAsyncCore(buffer, minimumBytes, throwOnEndOfStream: true, cancellationToken);
```
```suggestion
            return ReadAtLeastAsyncCore(buffer, minimumBytes, throwOnEndOfStream, cancellationToken);
```

eerhardt replied:
Oops - copy and paste error. Thanks for the catch. I'm writing unit tests now, I'll make sure I add a test for this too.

eerhardt committed:
Fix build and copy-paste error

stephentoub commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipHelper.cs:
```diff
@@ -40,17 +40,10 @@ internal static Encoding GetEncoding(string text)
-
-                totalBytesRead += bytesRead;
-                bytesLeftToRead -= bytesRead;
+                throw new IOException(SR.UnexpectedEndOfStream);
```
We don't want to use ReadExactly here because we want a different exception message?  Otherwise it would throw an EndOfStreamException which derives from IOException, right?

eerhardt replied:
Correct. Both the exception message and Type would be changed if we used ReadExactly here. I didn't want to change that behavior.

stephentoub commented on src/libraries/System.Net.Http/tests/FunctionalTests/LoopbackSocksServer.cs:
```diff
@@ -290,16 +290,9 @@ void HandleExceptions(Exception ex)
         }
 
-        private async ValueTask ReadToFillAsync(Stream stream, Memory<byte> buffer)
+        private ValueTask ReadToFillAsync(Stream stream, Memory<byte> buffer)
```
Can we just delete this helper and have the call sites use ReadExactlyAsync directly?

eerhardt replied:
I can. It will cause a few more changes in the file, it is called 10 times. But it will clean the code up a bit.

stephentoub commented on src/libraries/System.Net.Security/src/System/Net/StreamFramer.cs:
```diff
@@ -31,24 +31,17 @@ internal sealed class StreamFramer
-            int bytesRead;
-            int offset = 0;
-            while (offset < buffer.Length)
+            int bytesRead = await TAdapter.ReadAtLeastAsync(
```
There is a risk here.  Previously if the async read pends, it would force the ReadMessageAsync state machine to be allocated.  Now if the async read pends, it will force the ReadMessageAsync state machine to be allocated _and_ the ReadAtLeastAsync state machine to be allocated.

We should consider using the pooled async state machine builders on the ReadAtLeastAsync async methods.

eerhardt replied:
Is this as simple as adding

```C#
[AsyncMethodBuilder(typeof(PoolingAsyncValueTaskMethodBuilder))]
```

to the `Stream.ReadAtLeastAsyncCore` method?  Or is it more involved than that?

stephentoub replied:
Implementation-wise, that's it, easy peasy.  The important thing is validating performance.  The concern with such pooling is that it can have a negative impact on the larger system, especially if it ends up creating more gen2 to gen0 references (by referencing gen0 objects from pooled objects that are likely to eventually be gen2).  But I worry that without doing this, the methods will end up not being used because folks will see the allocation and choose to avoid the helpers as a result.

eerhardt replied:
Let's chat tomorrow about how I can validate the performance here. I'd like to get your thoughts on what to try / look for.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -332,6 +332,48 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
         }
 
+        public async ValueTask ReadExactlyAsync(Memory<byte> buffer, CancellationToken cancellationToken = default) =>
+            _ = await ReadAtLeastAsyncCore(buffer, buffer.Length, throwOnEndOfStream: true, cancellationToken).ConfigureAwait(false);
```
This is going to force another state machine into existence for async waits, just to cast the `ValueTask<int>` to a `ValueTask`. We should probably change this to be more like:
```C#
public ValueTask ReadExactlyAsync(..)
{
    ValueTask<int> vt = ReadAtLeastAsyncCore(...);

    if (vt.IsCompletedSuccessfully)
        return default;

    return new ValueTask(vt.AsTask());
}
```
If ReadAtLeastAsyncCore doesn't end up being annotated to use a pooled builder, that shouldn't result in any additional allocation.  If it does end up using a pooled builder, that AsTask will allocate, which is unfortunate, and it'd be unfortunate to then try to work around that by making this method use a pooling builder as well... at which point the cheapest thing might be to just copy-and-paste the code from ReadAtLeastAsyncCore to here.  We unfortunately lack a general-purpose mechanism for freely converting any `ValueTask<T>` to a `ValueTask`.

stephentoub replied:
Actually... this is all in corelib.  We can cheat and use internals on `ValueTask<T>` to extract the backing object and use that to construct the `ValueTask`.

eerhardt replied:
I took a stab at writing it like this, but wasn't 100% sure I was doing it right. Let me know what you think.

eerhardt replied:
One question I have is do I need to access the Result of the underlying `ValueTask<int>` in order for it to be returned to the pool?

stephentoub replied:
> One question I have is do I need to access the Result of the underlying ValueTask<int> in order for it to be returned to the pool?

In general, yes.  But if you're transferring the internals of the `ValueTask<int>` over to the `ValueTask`, you definitely _don't_ want to do that.  It will be returned to the pool when the `ValueTask` is awaited by the consumer.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -332,6 +332,48 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
+
+        public async ValueTask ReadExactlyAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken = default)
+        {
+            ValidateBufferArguments(buffer, offset, count);
```
These argument exceptions are going to be stored into the returned task rather than propagated.  We generally try to avoid that for ArgumentXxException.

eerhardt committed:
Fix WebSockets tests

eerhardt committed:
Add ReadAtLeast and ReadExactly unit tests

eerhardt committed:
Fix System.Net.Security tests

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -701,6 +743,48 @@ public virtual int ReadByte()
         }
 
+        public void ReadExactly(Span<byte> buffer) =>
+            _ = ReadAtLeastCore(buffer, buffer.Length, throwOnEndOfStream: true);
```
Both `ReadAtLeast` and `ReadExactly` can never read more bytes than the buffer that is passed into it.

So if you have a `buffer = new byte[10]`, and call `ReadAtLeast(buffer, 5)`, the Read call will read between `5-10` bytes of data. It is guaranteed to be `minimumBytes <= numRead` and `numRead <= buffer.Length`.

So yes, if you had "real data" in the buffer beyond the `minimumBytes` index, it is possible for that data to be overwritten by the Read call. You would need to slice the buffer down in this situation.  The same goes for `Read(Span<byte>)`, it can read to fill up the entire buffer.

eerhardt committed:
Respond to PR feedback.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -332,14 +333,36 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
+            if (vt.IsCompletedSuccessfully)
+                return default;
+
+            // use the ValueTask<int>'s backing object to create a ValueTask without allocating here.
+            object? obj = vt._obj;
+            Debug.Assert(obj is Task || obj is IValueTaskSource);
+
+            return obj is Task task ?
+                new ValueTask(task) :
+                new ValueTask((IValueTaskSource)obj, vt._token);
```
I'd suggest instead adding an `internal ValueTask DangerousAsValueTask()` method to `ValueTask<T>`.  Its implementation would be something like:
```C#
internal ValueTask DangerousAsValueTask()
{
    Debug.Assert(_obj is Task or IValueTaskSource);
    return new ValueTask(_object, _token, _continueOnCapturedContext);
}
```
and then this code in ReadExactlyAsync can just be:
```C#
return ReadAtLeastAsyncCore(...).DangerousAsValueTask();
```
...or something along those lines.

eerhardt replied:
I ended up creating a static `ValueTask.DangerousCreateFromTypedValueTask<TResult>` method because the ValueTask ctor that takes `object, short, bool` is `private`, and I didn't feel comfortable making it `internal` for all of CoreLib to see, and use by mistake. Thoughts?

stephentoub replied:
sounds good

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -701,6 +767,48 @@ public virtual int ReadByte()
+                {
+                    if (throwOnEndOfStream)
+                    {
+                        ThrowHelper.ThrowEndOfFileException();
```
I wonder why the method is called `ThrowEndOfFileException`, yet it throws an `EndOfStreamException` internally.

BTW, this is unrelated to your changes, but TIL of the existence of `ThrowHelper`. But it is in S.P.CoreLib, I would've liked to use it inside the new tar assembly. Would it make sense to request moving it to libraries/Common for wider usage? Or is it in S.P.CoreLib for optimization purposes I am not aware of?

eerhardt replied:
> I wonder why the method is called ThrowEndOfFileException, yet it throws an EndOfStreamException internally.

Agreed. It is a bit weird, but I wasn't going to mess with it here.

> But it is in S.P.CoreLib, I would've liked to use it inside the new tar assembly. Would it make sense to request moving it to libraries/Common for wider usage? Or is it in S.P.CoreLib for optimization purposes I am not aware of?

We have a few of them in across different assemblies. Since each assembly typically has its own exceptions to be thrown, I'm not sure it makes sense to reuse them across assemblies. We have one common (partial) one for ArgNull:

https://github.com/dotnet/runtime/blob/20cd393c1e577f0ff53cdc3a00edc3bcfc68ab7f/src/libraries/Common/src/System/ThrowHelper.cs#L12

See https://source.dot.net/#q=ThrowHelper for all of them.

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -701,6 +767,48 @@ public virtual int ReadByte()
+                    {
+                        ThrowHelper.ThrowEndOfFileException();
+                    }
+                    else
+                    {
+                        return totalRead;
+                    }
```
nit - You could get rid of the `else`.
```suggestion
                    return totalRead;
```

eerhardt replied:
I missed this in my current changes, but I'll fix it the next time I need to make changes.

carlossanlop commented on src/libraries/System.IO/tests/Stream/Stream.ReadExactly.cs:
```diff
@@ -0,0 +1,300 @@
+using System.Buffers;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Frameworks;
```
Unintended using?

eerhardt replied:
Yes, and I can't for the life of me figure out how it got in here. It must have been an automatic "add using" from VS as I was searching for a Type while writing code.

My next question was: "why are the tests compiling on my machine?". And discovered that the tests do reference NuGet.Frameworks locally through a `<PackageReference Include="Microsoft.NET.Test.Sdk"`. However, we turn that PackageReference off in CI, which is causing it to work locally, but fail in CI:

https://github.com/dotnet/runtime/blob/951df04d97405ff6dd13e708873549fb9acf9a6c/eng/testing/xunit/xunit.props#L17-L19

carlossanlop commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/StreamExtensions.cs:
```diff
@@ -78,6 +78,9 @@ internal static int TryReadAll(this Stream stream, byte[] buffer, int offset, in
 
 #if NETCOREAPP
         internal static int TryReadAll(this Stream stream, Span<byte> buffer)
+#if NET7_0_OR_GREATER
```
For my own education: Why is this specific case guarded by this preprocessor directive?

eerhardt replied:
I waffled on whether I should change this code or not. All the rest of the changes are to code that only ship in-box in the shared framework. `System.Reflection.Metadata` also ships in an OOB package, thus it builds for `net6.0` and `net7.0`. These new Stream APIs are only available in `net7.0`, and this code needs to handle both. So it needs to take separate paths.

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -757,6 +865,20 @@ protected static void ValidateBufferArguments(byte[] buffer, int offset, int cou
+
+            if (bufferLength < minimumBytes)
+            {
+                // TODO: pick the right exception message here
```
This one seems the closest one to what you need:
```cs
case ExceptionResource.ArgumentOutOfRange_SmallCapacity:
                    return SR.ArgumentOutOfRange_SmallCapacity;
```

The string is: `capacity was less than the current size`. But maybe we should add a new one if this is not clear enough. Also, it's weird to see that this message starts with lowercase.

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -757,6 +865,20 @@ protected static void ValidateBufferArguments(byte[] buffer, int offset, int cou
+            if (bufferLength < minimumBytes)
+            {
+                // TODO: pick the right exception message here
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.minimumBytes);
```
Is `ArgumentOutOfRangeException` the most appropriate for this check?

eerhardt replied:
I _think_ so. This is the exception that is thrown directly above this method:

https://github.com/dotnet/runtime/blob/20cd393c1e577f0ff53cdc3a00edc3bcfc68ab7f/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L754-L757

in this case, `minimumBytes` is "out of the valid range" of `0 - buffer.Length`.

carlossanlop reviewed (commented):
Thanks for this change! Left some comments and questions.

eerhardt committed:
Fix test build in CI

eerhardt committed:
Respond to PR feedback

eerhardt committed:
Fix System.Xml.Tests

eerhardt committed:
Add XML docs to the new APIs

eerhardt committed:
Add message for new exception.

eerhardt commented:
Thanks for the review so far. I now believe this is fully ready for review. PTAL.

eerhardt commented on src/libraries/System.IO/tests/Stream/Stream.ReadAtLeast.cs:
```diff
@@ -0,0 +1,274 @@
+            readInvokedCount = 0;
+            byte[] emptyBuffer = Array.Empty<byte>();
+
+            // The buffer is empty, so the Stream can only return 0 from Read. This looks like an end-of-stream,
```
Thoughts on this scenario? What should happen when an empty buffer is passed? Should we throw an ArgumentException instead? We could avoid the up front check if we only checked if the buffer was empty after Read returns `0`

adamsitnik replied:
iirc Sockets allow the users to perform a blocking zero-byte read which returns when data becomes available for reading. But I don't think that these new methods should be used for such scenarios. Even exactly the opposite, they should not be used.

My personal preference would be to throw for empty buffers, wait for users feedback and eventually change it before we ship .NET 7 RTM based on feedback.

eerhardt replied:
I changed it to throw a better exception in these cases.

davidfowl commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -332,6 +332,113 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
+        }
+
+        // No argument checking is done here. It is up to the caller.
+        [AsyncMethodBuilder(typeof(PoolingAsyncValueTaskMethodBuilder<>))]
```
This is what I came to check 👍🏾

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -701,6 +808,95 @@ public virtual int ReadByte()
+            int totalRead = 0;
+            do
+            {
+                int read = Read(buffer.Slice(totalRead));
```
The default implementation of `Stream.Read(Span<byte>)` is expensive, as it rents an array from array pool, reads the data into it using the `Read(array)`, copies the content to the span and returns the array to the pool:

https://github.com/dotnet/runtime/blob/65067052e433eda400c5e7cc9f7b21c84640f901/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L677-L695

In BCL all Streams override this method and provide optimal implementation, but it's not guaranteed for other streams.

@eerhardt @stephentoub I am not suggesting that we should have two implementations (one for methods that accept arrays, another for spans), but I would like to hear your opinion on that. Should we care about streams that don't override `Read(Span)` and `Write(Span)`?

stephentoub replied:
We use the span/memory-based overloads throughout all of the core library routines that read/write from/to streams.  If streams don't override these methods, they're already going to be deficient compared to what they could be; I don't think adding one more case here significantly moves the needle.

There were a few folks (not including me) in the API review discussion who felt it was important for usability that we expose a byte[]-based overload, but I don't think that means we need to duplicate our code to accommodate it.  If we didn't expose a byte[]-based overload, all use would end up going through the span-based path anyway.

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -701,6 +808,95 @@ public virtual int ReadByte()
+
+            return totalRead;
+        }
+
```
I love the fact that the new methods are not virtual. I owe a beer to everyone who pushed for that at the API review 🥇

adamsitnik commented on src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs:
```diff
@@ -1714,13 +1714,8 @@ internal static XmlReader CreateSqlReader(Stream input, XmlReaderSettings? setti
-                byteCount += read;
-            } while (read > 0 && byteCount < 2);
+            int bytesToRead = Math.Min(bytes.Length, 2);
+            int byteCount = input.ReadAtLeast(bytes, bytesToRead, throwOnEndOfStream: false);
```
the previous logic was hard to understand: read at least 2 bytes? now it looks much better 👍

adamsitnik commented on src/libraries/System.Runtime/ref/System.Runtime.cs:
```diff
@@ -9169,7 +9169,13 @@ public abstract partial class Stream : System.MarshalByRefObject, System.IAsyncD
+        public int ReadAtLeast(System.Span<byte> buffer, int minimumBytes, bool throwOnEndOfStream = true) { throw null; }
+        public System.Threading.Tasks.ValueTask<int> ReadAtLeastAsync(System.Memory<byte> buffer, int minimumBytes, bool throwOnEndOfStream = true, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
         public virtual int ReadByte() { throw null; }
+        public void ReadExactly(byte[] buffer, int offset, int count) { }
+        public void ReadExactly(System.Span<byte> buffer) { }
```
if we don't provide an implementation that is optimal for arrays I don't see why we should have two overloads for `ReadExactly` and only one for `ReadAtLeast`?

adamsitnik reviewed (commented):
Overall LGTM, but I would like to discuss two things:

* how should empty buffers be treated?
* should we optimize `Stream.ReadExactly(array, offset, count)` to use `Stream.Read(array)` or perhaps even remove this overload?

Big thanks for finding all the places where these methods could be used! It removed a lot of code duplication

stephentoub commented:
> My personal preference would be to throw for empty buffers

More specifically, I think we should throw for minimumBytes <= 0.  In the case of ReadExactly, that would mean an empty buffer.  In the case of ReadAtLeast, the buffer might not be empty.

There are two main situations where someone might be passing 0 to a read method:
- Want the "wait for data to be available" behavior
- Doing math to compute how much to read and not special-casing 0

Read is available for the former.  For the latter, a direct call to ReadAtLeast with a specified lower bound of 0 doesn't seem likely, given that this is expected to be used in scenarios like protocol parsing where you're expecting some number of bytes necessary to make forward progress, and for ReadExactly, it'll typically be used in situations where you're trying to read everything in the payload into some buffer you have... doing math is much more common with Read, where you might need to loop, and we wouldn't expect much looping around ReadExactly.

adamsitnik reviewed (approved):
Based on responses provided by @stephentoub it LGTM. Thanks!

eerhardt committed:
PR feedback

ReadAtLeast throws when minimumBytes is not positive.
ReadExactly throws when an empty buffer is passed, or count is not positive.

eerhardt committed:
Fix code to check for ReadAtLeast(0)

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -332,6 +332,120 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>A task that represents the asynchronous read operation.</returns>
+        /// <exception cref="ArgumentException">
+        /// <paramref name="buffer"/> was an empty buffer.
```
```suggestion
        /// <paramref name="buffer"/> is empty.
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -332,6 +332,120 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
+        /// <paramref name="buffer"/> was an empty buffer.
+        /// </exception>
+        /// <exception cref="EndOfStreamException">
+        /// Thrown when the end of the stream is reached before filling the <paramref name="buffer"/>.
```
```suggestion
        /// The end of the stream is reached before filling the <paramref name="buffer"/>.
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -332,6 +332,120 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
+        /// <param name="count">The number of bytes to be read from the current stream.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>A task that represents the asynchronous read operation.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="buffer"/> is null.</exception>
```
```suggestion
        /// <exception cref="ArgumentNullException"><paramref name="buffer"/> is <see langword="null"/>.</exception>
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -332,6 +332,120 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
+        /// <paramref name="offset"/> and <paramref name="count"/> exceed the length of <paramref name="buffer"/>.
+        /// </exception>
+        /// <exception cref="EndOfStreamException">
+        /// Thrown when the end of the stream is reached before reading <paramref name="count"/> number of bytes.
```
```suggestion
        /// The end of the stream is reached before reading <paramref name="count"/> number of bytes.
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -332,6 +332,120 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
+        /// <returns>A task that represents the asynchronous read operation.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="buffer"/> is null.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="offset"/> was outside the bounds of <paramref name="buffer"/>, or
+        /// <paramref name="count"/> was not a positive value, or the range specified by the combination of
+        /// <paramref name="offset"/> and <paramref name="count"/> exceed the length of <paramref name="buffer"/>.
```
```suggestion
        /// <paramref name="offset"/> is outside the bounds of <paramref name="buffer"/>.
        /// -or-
        /// <paramref name="count"/> is not a positive value.
        /// -or-
        /// The range specified by the combination of <paramref name="offset"/> and <paramref name="count"/> exceeds the length of <paramref name="buffer"/>.
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -332,6 +332,120 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
+        /// </param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>
+        /// A task that represents the asynchronous read operation. The value of its Result property contains the total number of bytes read into
```
I think this would be better for the user to be able to click on it.
```suggestion
        /// A task that represents the asynchronous read operation. The value of its <see cref="ValueTask{TResult}.Result"/> property contains the total number of bytes read into
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -332,6 +332,120 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
+        /// that many bytes are not currently available.
+        /// </returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="minimumBytes"/> was not a positive value, or greater than the length of <paramref name="buffer"/>.
```
```suggestion
        /// <paramref name="minimumBytes"/> is not a positive value, or is greater than the length of <paramref name="buffer"/>.
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -332,6 +332,120 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
+        /// <paramref name="minimumBytes"/> was not a positive value, or greater than the length of <paramref name="buffer"/>.
+        /// </exception>
+        /// <exception cref="EndOfStreamException">
+        /// Thrown when <paramref name="throwOnEndOfStream"/> is <see langword="true"/> and the end of the stream is reached before reading
+        /// <paramref name="minimumBytes"/> of data.
```
```suggestion
        /// <paramref name="throwOnEndOfStream"/> is <see langword="true"/> and the end of the stream is reached before reading
        /// <paramref name="minimumBytes"/> bytes of data.
```

eerhardt replied:
Does `minimumBytes bytes of data.` make sense with the redundancy?

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -701,6 +815,104 @@ public virtual int ReadByte()
+        /// </summary>
+        /// <param name="buffer">A region of memory. When this method returns, the contents of this region are replaced by the bytes read from the current stream.</param>
+        /// <exception cref="ArgumentException">
+        /// <paramref name="buffer"/> was an empty buffer.
```
```suggestion
        /// <paramref name="buffer"/> is empty.
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -701,6 +815,104 @@ public virtual int ReadByte()
+        /// <paramref name="buffer"/> was an empty buffer.
+        /// </exception>
+        /// <exception cref="EndOfStreamException">
+        /// Thrown when the end of the stream is reached before filling the <paramref name="buffer"/>.
```
```suggestion
        /// The end of the stream is reached before filling the <paramref name="buffer"/>.
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -701,6 +815,104 @@ public virtual int ReadByte()
+        /// <summary>
+        /// Reads <paramref name="count"/> number of bytes from the current stream and advances the position within the stream.
+        /// </summary>
+        /// <param name="buffer">An array of bytes. When this method returns, the buffer contains the specified byte array with the values between offset and (offset + count - 1) replaced by the bytes read from the current stream.</param>
```
Usually I don't mind about the length of the comment lines, but to keep it consistent with the rest of your comments, I suggest we split this.

```suggestion
        /// <param name="buffer">An array of bytes. When this method returns, the buffer contains the specified byte array with the values
        /// between <paramref name="offset"/> and (<paramref name="offset"/> + <paramref name="count"/> - 1) replaced by the bytes read from the current stream.</param>
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -701,6 +815,104 @@ public virtual int ReadByte()
+        /// <param name="buffer">An array of bytes. When this method returns, the buffer contains the specified byte array with the values between offset and (offset + count - 1) replaced by the bytes read from the current stream.</param>
+        /// <param name="offset">The byte offset in <paramref name="buffer"/> at which to begin storing the data read from the current stream.</param>
+        /// <param name="count">The number of bytes to be read from the current stream.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="buffer"/> is null.</exception>
```
```suggestion
        /// <exception cref="ArgumentNullException"><paramref name="buffer"/> is <see langword="null"/>.</exception>
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -701,6 +815,104 @@ public virtual int ReadByte()
+        /// <param name="count">The number of bytes to be read from the current stream.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="buffer"/> is null.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="offset"/> was outside the bounds of <paramref name="buffer"/>, or
+        /// <paramref name="count"/> was not a positive value, or the range specified by the combination of
+        /// <paramref name="offset"/> and <paramref name="count"/> exceed the length of <paramref name="buffer"/>.
```
```suggestion
        /// <paramref name="offset"/> is outside the bounds of <paramref name="buffer"/>.
        /// -or-
        /// <paramref name="count"/> is not a positive value.
        /// -or-
        /// The range specified by the combination of <paramref name="offset"/> and <paramref name="count"/> exceeds the length of <paramref name="buffer"/>.
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -701,6 +815,104 @@ public virtual int ReadByte()
+        /// <paramref name="offset"/> and <paramref name="count"/> exceed the length of <paramref name="buffer"/>.
+        /// </exception>
+        /// <exception cref="EndOfStreamException">
+        /// Thrown when the end of the stream is reached before reading <paramref name="count"/> number of bytes.
```
```suggestion
        /// The end of the stream is reached before reading <paramref name="count"/> number of bytes.
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -701,6 +815,104 @@ public virtual int ReadByte()
+        /// of bytes allocated in the buffer if that many bytes are not currently available.
+        /// </returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="minimumBytes"/> was not a positive value, or greater than the length of <paramref name="buffer"/>.
```
```suggestion
        /// <paramref name="minimumBytes"/> is not a positive value, or is greater than the length of <paramref name="buffer"/>.
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:
```diff
@@ -701,6 +815,104 @@ public virtual int ReadByte()
+        /// <paramref name="minimumBytes"/> was not a positive value, or greater than the length of <paramref name="buffer"/>.
+        /// </exception>
+        /// <exception cref="EndOfStreamException">
+        /// Thrown when <paramref name="throwOnEndOfStream"/> is <see langword="true"/> and the end of the stream is reached before reading
```
```suggestion
        /// <paramref name="throwOnEndOfStream"/> is <see langword="true"/> and the end of the stream is reached before reading
```

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs:
```diff
@@ -398,6 +398,15 @@ internal void ThrowIfCompletedUnsuccessfully()
             }
         }
 
+        /// <summary>Transfers the <see cref="ValueTask{TResult}"/> to a <see cref="ValueTask"/> instance.</summary>
+        /// <remarks>The <see cref="ValueTask{TResult}"/> should not be used after calling this method.</remarks>
```
This is an internal method. I'm ok with having triple slash comments, but putting that text in the remarks means it won't show up in intellisense. And since it's not ported to MS Docs (because it's internal), I suggest we put all the text in the summary.

```suggestion
        /// <summary>Transfers the <see cref="ValueTask{TResult}"/> to a <see cref="ValueTask"/> instance.
        /// The <see cref="ValueTask{TResult}"/> should not be used after calling this method.</summary>
```

carlossanlop reviewed (commented):
Left some documentation suggestions.

eerhardt committed:
Xml doc feedback

eerhardt committed:
Preserve behavior in StreamReader.FillBuffer when passed 0.

GSPP commented:
How would `ReadExactly` be used to read a length-prefixed binary value from some protocol? A zero-length binary data is no special case and should just work.

My understanding is that this call would currently throw. I'm not sure why there is a need to create a special case where, naturally, no special case is necessary.

stephentoub commented:
> I'm not sure why there is a need to create a special case

Because otherwise it could hang until there's data available. If you want that behavior, Read already exists and provides everything needed.

GSPP commented:
Oh, now I see. If we have a zero-length read, this would cause many streams to wait until more data is incoming! That's a bug, and I'm afraid I will now have to check some of my code for this particular mistake 😁. Thanks for explaining.

stephentoub commented:
> An alternative is special-casing the new methods to handle 0 by always returning immediately

@eerhardt, what would you think of this?

eerhardt commented:
I don't really like having different behavior between `Read(buffer, 0, 0)` and `ReadAtLeast`(buffer, 0)`. My first proposal was to make these behave roughly equivalent - block until data is available. Passing in an empty buffer into both `ReadAtLeast` and `ReadExactly` became a problem though, since `0` was returned from the `Read` operation, and that looked like an end-of-stream.

At a high-level, I can think of the following possibilities:

1. Check up front and throw an exception. (current PR behavior)
2. Return immediately, don't do a "read", even though a "read" method was called.
3. Always call `Read`, with 2 options on when an empty buffer is passed:
    a. Throw an exception
    b. Special case an empty buffer, and allow for `0` to be returned from `Read`. Return successfully for `ReadExactly` and `0` for ReadAtLeast.

My current thinking is that 3(b) seems to be the most consistent with `Stream.Read`, and would be reasonable behavior.

stephentoub commented:
> My current thinking is that 3(b) seems to be the most consistent with Stream.Read, and would be reasonable behavior.

I don't think that's desirable.  It means a scenario like:
```C#
int numBytesInPayload = ReadPayloadSize(stream);
byte[] payload = new byte[numBytesInPayload];
stream.ReadExactly(payload);
```
could block indefinitely.

I do think 0 is special for ReadExactly and ReadAtLeast and needs to be treated differently from what Read does.  We could choose to throw (1) or return immediately (2), but calling Read with 0 (3) is problematic.

stephentoub commented:
> Are there two different scenarios here?

That's part of my question... what is the scenario for calling ReadAtLeast with a non-zero-length buffer and minimumBytes == 0?  Why wouldn't you just use Read?

eerhardt commented:
Maybe the behavior split makes sense between the new methods then?

* ReadExactly with an empty buffer (or `count == 0` for the array overload) - no-ops
* ReadAtLeast with `minimumBytes == 0` - throws

stephentoub commented:
> I just wouldn't expect ReadAtLeast to no-op

Why?  It's in the name "at least"... when it has at least 0, it's done.

eerhardt committed:
Handle ReadExactly with an empty buffer, and ReadAtLeast with minimumBytes == 0.

Both of these cases are a no-op. No exception is thrown. A read won't be issued to the underlying stream. They won't block until data is available.
